controllers/drain_controller.go

```diff
@@ -0,0 +1,464 @@
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "node object doesn't exist", "nodeName", req.Name)
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "failed to get node object from api re-queue the request", "nodeName", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// get sriovNodeNodeState object
+	nodeNetworkState := &sriovnetworkv1.SriovNetworkNodeState{}
+	err = dr.Get(ctx, req.NamespacedName, nodeNetworkState)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			reqLogger.Error(err, "sriovNetworkNodeState object doesn't exist", "nodeName", req.Name)
+			return ctrl.Result{}, nil
+		}
+
+		reqLogger.Error(err, "failed to get sriovNetworkNodeState object from api re-queue the request", "nodeName", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// create the drain state annotation if it doesn't exist in the sriovNetworkNodeState object
+	nodeStateDrainAnnotationCurrent, NodeStateDrainAnnotationCurrentExist := nodeNetworkState.Annotations[constants.NodeStateDrainAnnotationCurrent]
+	if !NodeStateDrainAnnotationCurrentExist {
+		err = utils.AnnotateObject(nodeNetworkState, constants.NodeStateDrainAnnotationCurrent, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeStateDrainAnnotationCurrent = constants.DrainIdle
+	}
+
+	// create the drain state annotation if it doesn't exist in the node object
+	nodeDrainAnnotation, nodeDrainAnnotationExist := node.Annotations[constants.NodeDrainAnnotation]
+	if !nodeDrainAnnotationExist {
+		err = utils.AnnotateObject(node, constants.NodeDrainAnnotation, constants.DrainIdle, dr.Client)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		nodeDrainAnnotation = constants.DrainIdle
+	}
+
+	reqLogger.V(2).Info("Drain annotations", "nodeAnnotation", nodeDrainAnnotation, "nodeStateAnnotation", nodeStateDrainAnnotationCurrent)
+
+	// Check the node request
+	if nodeDrainAnnotation == constants.DrainIdle {
+		// this cover the case the node is on idle
+
+		// node request to be on idle and the currect state is idle
+		// we don't do anything
+		if nodeStateDrainAnnotationCurrent == constants.DrainIdle {
+			reqLogger.Info("node and nodeState are on idle nothing todo")
+			return reconcile.Result{}, nil
+		}
+
+		// we have two options here:
+		// 1. node request idle and the current status is drain complete
+		// this means the daemon finish is work, so we need to clean the drain
+		//
+		// 2. the operator is still draining the node but maybe the sriov policy changed and the daemon
+		//  doesn't need to drain anymore, so we can stop the drain
+		if nodeStateDrainAnnotationCurrent == constants.DrainComplete ||
+			nodeStateDrainAnnotationCurrent == constants.Draining {
+			completed, err := dr.drainer.CompleteDrainNode(ctx, node)
+			if err != nil {
+				reqLogger.Error(err, "failed to complete drain on node")
+				return ctrl.Result{}, err
+			}
+
+			// if we didn't manage to complete the un drain of the node we retry
```

Comment:
This would happen if something else changes on the node annotation we get called while we were still draining?